kornia/contrib/models/rt_detr/architecture/hybrid_encoder.py

```diff
@@ -0,0 +1,141 @@
+"""Based on the code from https://github.com/PaddlePaddle/PaddleDetection/blob/5d1f888362241790000950e2b63115dc8d1c
+6019/ppdet/modeling/transformers/hybrid_encoder.py."""
+
+from __future__ import annotations
+
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+from kornia.core import Module, Tensor, concatenate
+
+from .common import ConvNormAct
+
+
+# NOTE: conv2 can be fused into conv1
+class RepVggBlock(Module):
+    def __init__(self, in_channels: int, out_channels: int):
+        super().__init__()
+        self.conv1 = ConvNormAct(in_channels, out_channels, 3, act="none")
+        self.conv2 = ConvNormAct(in_channels, out_channels, 1, act="none")
+
+    def forward(self, x: Tensor) -> Tensor:
+        return F.silu(self.conv1(x) + self.conv2(x))
+
+
+class CSPRepLayer(Module):
+    def __init__(self, in_channels: int, out_channels: int, num_blocks: int):
+        super().__init__()
+        self.conv1 = ConvNormAct(in_channels, out_channels, 1, act="silu")
+        self.conv2 = ConvNormAct(in_channels, out_channels, 1, act="silu")
+        self.bottlenecks = nn.Sequential(*[RepVggBlock(out_channels, out_channels) for _ in range(num_blocks)])
+
+    def forward(self, x: Tensor) -> Tensor:
+        return self.bottlenecks(self.conv1(x)) + self.conv2(x)
+
+
+# almost identical to nn.TransformerEncoderLayer
+# but add positional embeddings to q and k
+class AIFI(Module):
+    def __init__(self, embed_dim: int, num_heads: int, dim_feedforward: int, dropout: float = 0.0):
+        super().__init__()
+        self.self_attn = nn.MultiheadAttention(embed_dim, num_heads, dropout)  # NOTE: batch_first = False
+        self.dropout1 = nn.Dropout(dropout)
+        self.norm1 = nn.LayerNorm(embed_dim)
+
+        self.linear1 = nn.Linear(embed_dim, dim_feedforward)
+        self.act = nn.GELU()
+        self.dropout = nn.Dropout(dropout)
+        self.linear2 = nn.Linear(dim_feedforward, embed_dim)
+        self.dropout2 = nn.Dropout(dropout)
+        self.norm2 = nn.LayerNorm(embed_dim)
+
+    def forward(self, x: Tensor) -> Tensor:
+        # NOTE: cache build_2d_sincos_pos_emb to buffer, if input size is known?
+        # using post-norm
+        N, C, H, W = x.shape
+        x = x.permute(2, 3, 0, 1).flatten(0, 1)  # (N, C, H, W) -> (H * W, N, C)
+        q = k = x + self.build_2d_sincos_pos_emb(H, W, C)
+        x = self.norm1(x + self.dropout1(self.self_attn(q, k, x)[0]))
+        x = self.norm2(x + self.ffn(x))
+        x = x.view(H, W, N, C).permute(2, 3, 0, 1)  # (H * W, N, C) -> (N, C, H, W)
+        return x
+
+    def ffn(self, x: Tensor) -> Tensor:
+        return self.dropout2(self.linear2(self.dropout(self.act(self.linear1(x)))))
+
+    @staticmethod
+    def build_2d_sincos_pos_emb(w: int, h: int, embed_dim: int, temp: float = 10_000.0) -> Tensor:
```

Comment:
That could be a utils next to create_meshgrid useful for the transformers family

Reply:
For later thoug, add a todo note now